### PR TITLE
Do not allow operator! on target qubits.

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

File: test/AST-error/negation-error.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ ******************************************************************************/
+
+// RUN: cudaq-quake -verify %s
+
+#include <cudaq.h>
+
+struct NegationOperatorTest {
+  void operator()() __qpu__ {
+    cudaq::qreg qr(3);
+    x<cudaq::ctrl>(qr[0], qr[1], !qr[2]); // expected-error{{target qubit cannot be negated}}
```
Comment (bettinaheim) [inline test/AST-error/negation-error.cpp:16]: I may be good to add a test case for a custom kernel with cudaq::control as well. 
I'd also test an operation with more than one target qubit.

Comment (schweitzpgi) [reply test/AST-error/negation-error.cpp:16]: There are already tests for `!` used in a control position. Is that what you mean?

Comment (amccaskey) [reply test/AST-error/negation-error.cpp:16]: I think she means `cudaq::control(anotherKernel, {q[0], !q[1]}, q[2]);` I'm not sure if the bridge is setup for this yet, but we should do it.

Comment (bettinaheim) [reply test/AST-error/negation-error.cpp:16]: This is why I am asking. I don't think we should have a difference in operator support between builtin operations and custom kernels.

Comment (amccaskey) [reply test/AST-error/negation-error.cpp:16]: I agree with that. I think in the case of `cudaq::control` we could setup the bridge to allow negations on the control qubits, and then scan any of the trailing arguments, and if one is a qubit type, then disallow (throw an error) if anyone tries to negate it.

Comment (schweitzpgi) [reply test/AST-error/negation-error.cpp:16]: OK, if I'm understanding this, "custom" means user-defined device kernels.

We have a problem there in that user-defined device kernels do not declare which quantum parameters are controls and which are targets. In fact, a single quantum parameter can be both, either, or neither depending on dynamic behavior, control-flow, etc.

Comment (amccaskey) [reply test/AST-error/negation-error.cpp:16]: This is kind of what I meant when I said it was not quite as trivial. However in this case, I think we should assume that any qubit operands in the trailing arguments of `cudaq::control` are targets, and therefore cannot be negated. 
```cpp
__qpu__ void test(cudaq::qubit& q, cudaq::qubit& r, double d);
__qpu__ void hello() {
  qoda::qreg q(4);
  ...
  cudaq::control(test, {q[0], !q[1] /*should be ok*/}, q[2], !q[3] /*should not be ok*/, 2.2); 
```
Open to other thoughts on this though. @boschmitt

Comment (schweitzpgi) [reply test/AST-error/negation-error.cpp:16]: Makes sense insofar as an application of the `cudaq::control` kernel invocation.

Are we going to disallow `!q` as arguments to user-defined pure-device kernels?

Comment (schweitzpgi) [reply test/AST-error/negation-error.cpp:16]: Moving discussion to https://github.com/NVIDIA/cuda-quantum/issues/23

**Review by bettinaheim — APPROVED** [review]
Approving, since the additional test cases I requested require more work that will be processed in a separate PR. @schweitzpgi Please make sure there is accurate documentation for when and how negation can be used.

**Pull request merged**